opendb-core/src/main/java/org/openplacereviews/opendb/service/DBConsensusManager.java

```diff
@@ -909,17 +892,250 @@ public boolean validateExistingOperation(OpOperation op) {
+
+			int k = 7;
+			for (String id : deletedObjects.get(i)) {
+				args[k] = id;
+				k++;
+			}
+			insertBatch.add(args);
+		}
+
+		return insertBatch;
+	}
+
+	private void generateSignedByArguments(Object[] args, List<String> signedBy, int k) {
+		for (int i = 0; i < 2; i++) {
+			if (signedBy.size() > 1) {
+				args[k] = signedBy.get(i);
+			} else {
+				if (k == 6) {
+					args[k] = null;
+				} else {
+					args[k] = signedBy.get(0);
+				}
+			}
+			k++;
+		}
+	}
+
+	public void saveHistoryForObjects(OpOperation op, Date date, List<OpObject> newEditedObjects) {
+		int userAmountFields = 2, totalAmountFields = 12;
+
+		List<Object[]> allBatches = new LinkedList<>();
+		allBatches.addAll(prepareArgumentsForHistoryBatch(op.getCreated(), op, date, HistoryDTO.HistoryObject.Status.CREATED, totalAmountFields));
+		allBatches.addAll(prepareArgumentsForHistoryBatch(op.getDeleted(), op, date, totalAmountFields));
+		allBatches.addAll(prepareArgumentsForHistoryBatch(newEditedObjects, op, date, HistoryDTO.HistoryObject.Status.EDITED, totalAmountFields));
+
+		dbSchema.insertObjIntoHistoryTableBatch(allBatches, OP_OBJ_HISTORY_TABLE, jdbcTemplate, userAmountFields, totalAmountFields);
+
+	}
+
+	public HistoryDTO getHistoryForUser(List<String> user, int limit, String sortType) {
+		return jdbcTemplate.query("SELECT p1, p2, p3, p4, p5, time, obj, type, status from " + OP_OBJ_HISTORY_TABLE + " where " +
+						dbSchema.generatePKString(OP_OBJ_HISTORY_TABLE, "u%1$d = ?", " AND ", user.size()) +
+						" ORDER BY time,dbid " + sortType + " LIMIT " + limit, user.toArray(),
+				new ResultSetExtractor<HistoryDTO>() {
+					@Override
+					public HistoryDTO extractData(ResultSet rs) throws SQLException, DataAccessException {
+						HistoryDTO historyDTO = new HistoryDTO();
+						List<List<String>> objIds = new LinkedList<>();
+						List<HistoryDTO.HistoryObject> allObjects = new LinkedList<>();
+						while (rs.next()) {
+							HistoryDTO.HistoryObject historyObject = new HistoryDTO.HistoryObject();
+							List<String> ids = new ArrayList<>();
+							for (int i = 1; i <= 5; i++) {
+								if (rs.getString(i) != null) {
+									ids.add(rs.getString(i));
+								}
+							}
+							historyObject
+									.setUser(user)
+									.setId(ids)
+									.setDate(formatFullDate(rs.getTimestamp(6)))
+									.setOpObject(formatter.parseObject(rs.getString(7)))
+									.setType(rs.getString(8))
+									.setStatus(HistoryDTO.HistoryObject.Status.getStatus(rs.getInt(9)));
+
+							allObjects.add(historyObject);
+							if (!objIds.contains(ids)) {
+								objIds.add(ids);
+							}
+						}
+						Map<List<String>, List<HistoryDTO.HistoryObject>> history = new LinkedHashMap<>();
+						generateObjMapping(objIds, allObjects, history);
+
+						historyDTO.setObjects(history);
+
+						return historyDTO;
+					}
+				});
+	}
+
+	public HistoryDTO getHistoryForObject(List<String> id, int limit, String sortType) {
```

Comment:
where object type? i.e. sys.vote, sys.operation. Id is not part of type if I remember